src/util/manager/firestore.ts

```diff
@@ -2,16 +2,258 @@ import Bot from "../../api/bot";
 import { settings } from "../../settings";
 import { Firestore } from "@google-cloud/firestore";
 import Manager from "../../api/manager";
+import {
+  Circle,
+  CircleData,
+  Coper,
+  CoperData,
+  response,
+  ResponseType,
+  RRMessageData,
+  rrMessageData,
+  Task,
+} from "../../api/schema";
+import { z } from "zod";
+
+export const cacheTypes = z.object({
+  responses: z.map(z.string(), response),
+  rrmessages: z.map(z.string(), rrMessageData),
+  circles: z.map(z.string(), CircleData),
+  copers: z.map(z.string(), CoperData),
+});
+
+export type CacheTypes = z.infer<typeof cacheTypes>;
+
+export const schemaTypes = z.enum([
+  "responses",
+  "rrmessages",
+  "circle",
+  "coper",
+  "member",
+  "task",
+]);
+
+export type Schema = z.infer<typeof schemaTypes>;
 
 export default class FirestoreManager extends Manager {
-  public firestore: Firestore | undefined;
+  public firestore!: Firestore;
+  public cache: CacheTypes;
   constructor(bot: Bot) {
     super(bot);
+    this.cache = {
+      responses: new Map(),
+      rrmessages: new Map(),
+      circles: new Map(),
+      copers: new Map(),
+    };
   }
   async init() {
     this.firestore = new Firestore({
       projectId: settings.firestore.projectId,
       keyFilename: settings.firestore.keyFilename,
     });
+    await this.recache("responses", "responses");
+    await this.recache("rrmessages", "rrmessages");
+    await this.recache("circle", "circles");
+    await this.recache("coper", "copers");
+  }
+
+  private async recache(schema: Schema, cache?: keyof CacheTypes) {
+    try {
+      const data = await this.firestore.collection(schema).get();
+      if (cache) {
+        this.cache[cache] = new Map();
+        data.forEach((doc) => {
+          // this.cache[cache]?.set(doc.id, doc.data() as any);
+          if (cache === "circles") {
```

Comment:
Would add a comment here about how firebase stores dates and why its not compatible with default zod parsing